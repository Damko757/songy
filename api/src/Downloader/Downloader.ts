import ffmpeg from "fluent-ffmpeg";
import { PassThrough, Readable } from "stream";
import { Metadator } from "../Metadata/Metadator.js";
import type { YTDL_DownloadOptions } from "@ybd-project/ytdl-core";
import type { Metadata } from "../../../shared/Entities/Metadata/Metadata.js";

export class Downloader {
  link: string;
  metadator: Metadator;

  constructor(link: string) {
    this.link = link;
    this.metadator = new Metadator(this.link);
  }

  audioStream(
    options: YTDL_DownloadOptions & {
      bitrate?: number;
    } = {}
  ) {
    const self = this;
    const outStream = new PassThrough();

    this.metadator
      .rawMetaData()
      .then((info) => {
        self.metadator
          .getYtDlInstance()
          .downloadFromInfo(info, { quality: "highestaudio", ...options })
          .then((stream) => {
            const reader = stream.getReader();
            ffmpeg()
              .input(
                new Readable({
                  async read() {
                    // Function to read the chunks asynchronously
                    const pushChunk = async () => {
                      try {
                        const { done, value } = await reader.read();

                        if (done) {
                          this.push(null); // No more data, signal end of stream
                        } else {
                          this.push(value); // Push the chunk of data to the Node.js stream
                          pushChunk(); // Continue reading
                        }
                      } catch (err) {
                        this.emit("error", err); // Handle any errors
                      }
                    };

                    pushChunk(); // Start reading the chunks
                  },
                })
              )
              .audioBitrate(options.bitrate ?? 320)
              .format("mp3")
              .pipe(outStream);
          })
          .catch((e) => {
            console.error(e);
            outStream.emit("error", e);
          });
      })
      .catch((e) => {
        outStream.emit("error", e);
      });

    return outStream;
  }
}
