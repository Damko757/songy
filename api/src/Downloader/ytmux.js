
// require all the things!
const fs = require("fs");
const ytdl = require('@distube/ytdl-core');
const ffmpegPath = require('ffmpeg-static');
const cp = require('child_process');
const stream = require('stream');

// default export: the ffmpeg muxer
const ytmux = (link, options = {}) => {
    const result = new stream.PassThrough({ highWaterMark: options.highWaterMark || 1024 * 512 });
    console.info("Loading:", link);
    ytdl.getInfo(link).then(info => {
      console.info(info)
      ytdl.downloadFromInfo(info, { quality: 'highestaudio', ...options }).pipe(fs.createWriteStream("out/vidOnly.mp4"));
      ytdl.downloadFromInfo(info, { quality: 'highestaudio', ...options }).pipe(fs.createWriteStream("out/audioOnly.mp4"));
      //ffmpeg -fflags +igndts -i video.mp4 -i audio.mp4 -c:v copy -c:a aac output.mp4 
        audioStream = ytdl.downloadFromInfo(info, { quality: 'highestaudio', ...options });
        videoStream = ytdl.downloadFromInfo(info, { quality: 'highestvideo', ...options });
        videoStream.on('progress', (...args) => result.emit('progress', ...args));
        videoStream.on("finish", (...args) => console.log("FINISH", ...args));
        audioStream.on("finish", (...args) => console.log("FINISH", ...args));
        // create the ffmpeg process for muxing
        ffmpegProcess = cp.spawn(ffmpegPath, [
            // supress non-crucial messages
            // '-loglevel', '8', '-hide_banner',
            "-fflags", "+igndts",
            // input audio and video by pipe
            '-i', 'pipe:3', //'-i', 'pipe:4',
            // map audio and video correspondingly
            '-map', '0:a', //'-map', '1:v',
            // no need to change the codec
            '-c', 'copy',
            // output mp4 and pipe
            '-f', 'matroska', 'pipe:5'
        ], {
            // no popup window for Windows users
            windowsHide: true,
            stdio: [
                // silence stdin/out, forward stderr,
                'inherit', 'inherit', 'inherit',
                // and pipe audio, video, output
                'pipe', 'pipe', 'pipe'
            ]
        });
        audioStream.pipe(ffmpegProcess.stdio[3]);
        videoStream.pipe(ffmpegProcess.stdio[4]);
        // ffmpegProcess.stdio[5].pipe(result);
        ffmpegProcess.stdio[5].pipe(fs.createWriteStream("out/full.mp4"));
    }).catch(err => result.emit('error', err));
    return result;
};

// export it
module.exports = ytmux;

// export other functions, in case you want them
ytmux.download = ytdl;
ytmux.chooseFormat = ytdl.chooseFormat;
ytmux.downloadFromInfo = ytdl.downloadFromInfo;
ytmux.filterFormats = ytdl.filterFormats;
ytmux.getBasicInfo = ytdl.getBasicInfo;
ytmux.getInfo = ytdl.getInfo;
ytmux.getURLVideoID = ytdl.getURLVideoID;
ytmux.getVideoID = ytdl.getVideoID;
ytmux.validateID = ytdl.validateID;
ytmux.validateURL = ytdl.validateURL;