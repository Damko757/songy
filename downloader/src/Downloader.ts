import { PassThrough, Readable, Writable } from "stream";
import fs from "fs";
import cp, { exec, spawn } from "child_process";
import ytdl, { type downloadOptions } from "@distube/ytdl-core";
import { ObjectId } from "mongoose";
import path, { dirname } from "path";
import { fileURLToPath } from "url";
import { DownloadJob } from "./Commands/Command.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

export class Downloader {
  link: string;

  static supportedExtensions: Record<"audio" | "video", string[]> = {
    audio: ["mp3"],
    video: ["mp4"],
  };
  static downloadDirectory = path.resolve(
    __dirname,
    "..",
    "..",
    "..",
    "..",
    "out"
  );

  constructor(link: string) {
    this.link = link;
  }

  /**
   * Resolves download path
   * @param id Id of file to download
   * @param extension Extension of file. Can be undefined, then no extension is added
   * @param extra Additional value, if temporary file is created (e. g. audio and video-only files, then merge)
   */
  static downloadPath(
    id: ObjectId | string,
    extension?: DownloadJob["extension"],
    extra?: string
  ) {
    if (!fs.existsSync(Downloader.downloadDirectory))
      fs.mkdirSync(Downloader.downloadDirectory);

    // TODO: Fix for special directory
    return path.resolve(
      Downloader.downloadDirectory,
      `${id}${extra ? "_" + extra : ""}${extension ? "." + extension : ""}`
    );
  }

  /**
   * Fetches info for `this.link`
   * @param options Get info options for ytdl-core
   * @returns Information for provided link
   */
  getInfo(options: ytdl.getInfoOptions = {}) {
    return ytdl.getInfo(this.link, options);
  }

  /**
   * Validates Youtube URL/video ID
   * @returns bool
   */
  isValid() {
    return this.link.includes("?")
      ? ytdl.validateURL(this.link)
      : ytdl.validateID(this.link);
  }

  /**
   * Streams audio only
   * Original stream is converted via FFMPEG. Progress and info event are added as an extra
   * @param options
   * @returns PassThrough audio stream
   */
  audioStream(options: ytdl.downloadOptions & { bitrate?: number } = {}) {
    return new Promise<Readable>(async (resolve, reject) => {
      ytdl
        .getInfo(this.link)
        .then((info) => {
          const format =
            options.format ??
            ytdl.chooseFormat(info.formats, {
              quality: options.quality ?? "highestaudio",
              ...options,
            }); // Fetching format for FFMPEG

          const videoStream = ytdl.downloadFromInfo(info, {
            quality: options.quality ?? "highestaudio",
            filter: "audio",
            ...options,
          });

          const audioStream = new PassThrough();

          // Emitting Ytdl-core events
          videoStream.on("progress", (...args) =>
            audioStream.emit("progress", ...args)
          );
          videoStream.on("info", (...args) => {
            audioStream.emit("info", ...args);
          });

          const ffmpeg = spawn("ffmpeg", [
            "-i",
            "pipe:0", // Input from stdin
            "-vn", // No video
            ...(format.audioBitrate
              ? ["-b:a", format.audioBitrate.toString() + "k"]
              : []), // Audio bitrate from format
            "-acodec",
            "libmp3lame", // Audio codec
            "-f",
            "mp3", // Output format
            "pipe:1", // Output to stdout
          ]);

          // Pipe video input into FFmpeg
          videoStream.pipe(ffmpeg.stdin);

          // Pipe FFmpeg output to audioOutput stream
          ffmpeg.stdout.pipe(audioStream);

          ffmpeg.on("data", console.error);

          resolve(audioStream);
        })
        .catch(reject);
    });
  }

  /**
   * Download video Only
   * @param options
   */
  videoStream(options: downloadOptions = {}) {
    return new Promise<Readable>((resolve, reject) => {
      ytdl
        .getInfo(this.link)
        .then((info) => {
          const videoStream = ytdl.downloadFromInfo(info, {
            filter: "video",
            ...options,
          });
          return resolve(videoStream);
        })
        .catch(reject);
    });
  }

  /**
   * Saves `stream` into file `filename`
   * @param stream Readable stream to save
   * @param filename Of saved stream
   * @returns Promise resolved with filename upon completed write
   */
  async saveStream(stream: Readable, filename: string) {
    return new Promise<string>((resolve, reject) => {
      console.info("Saving to:", filename);
      const fileStream = fs.createWriteStream(filename);
      fileStream.on("error", reject);
      fileStream.on("close", () => resolve(filename));
      stream.pipe(fileStream);
    });
  }

  /**
   * Combines video and audio files into single MP4 file
   * @param videoFilename Filename of video file
   * @param audioFilename Filename of audio file
   * @param outFilename Filename of resulting video file
   * @returns Promise resolved upon completed combining
   */
  createCombinedVideoAudio(
    videoFilename: string,
    audioFilename: string,
    outFilename: string
  ) {
    return new Promise<void>((resolve, reject) => {
      const cmd = `ffmpeg -i ${videoFilename} -i ${audioFilename} -c:v copy -c:a aac -shortest ${outFilename} -y`;

      exec(cmd, (error, stdout, stderr) => {
        if (error) {
          console.error(`Error: ${error.message}`);
          reject(error);
          return;
        }
        if (stderr) {
          console.error(`FFmpeg stderr: ${stderr}`);
        }
        console.log("FFmpeg output:", stdout);
        resolve();
      });
    });
  }
}
