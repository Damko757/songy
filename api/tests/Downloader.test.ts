import { expect, describe, it, beforeEach } from "bun:test";
import { Downloader } from "../src/Downloader/Downloader";
import fs from "fs";
import path, { resolve } from "path";
import sha256 from "crypto-js/sha256";

describe("FFMPEG", () => {
  beforeEach(() => {
    fs.readdir("./out", (err, files) => {
      if (err) throw err;
      for (const file of files) {
        if (file.endsWith(".jpg")) continue;
        fs.unlink(path.join("./out", file), (err) => {
          if (err) throw err;
        });
      }
    });
  });

  describe.only("MP3", () => {
    it("Audio only", async () => {
      // expect.assertions(1);
      const downloader = new Downloader(
        "33fPaNWvyzE" //https://www.youtube.com/watch?v=
      );
      const audioStream = downloader.audioStream();
      const buf: Uint8Array[] = [];
      audioStream.on("data", (chunk) => {
        buf.push(chunk);
      });

      await new Promise((resolve, reject) => {
        audioStream.on("end", () => {
          expect(sha256(Buffer.concat(buf).join(""))).toMatchSnapshot();
          fs.writeFileSync("out/duck.mp3", Buffer.concat(buf));
          resolve(true);
        });

        audioStream.on("error", (e) => {
          reject(e);
        });
      });
    });
    it.only(
      "Audio with Metadata",
      async () => {
        const downloader = new Downloader("sduDiIGqvfQ");
        const metas = await downloader.metadator.metaDatas();
        const audioStream = downloader.audioStream({
          metadata: metas.spotify?.[0] ?? {},
          bitrate: 320,
        });

        const buf: Uint8Array[] = [];
        audioStream.on("data", (chunk) => {
          buf.push(chunk);
        });

        await new Promise((resolve, reject) => {
          audioStream.on("end", () => {
            expect(sha256(Buffer.concat(buf).join(""))).toMatchSnapshot();
            fs.writeFileSync("out/carpet.mp3", Buffer.concat(buf));
            resolve(true);
          });

          audioStream.on("error", (e) => {
            reject(e);
          });
        });
      },
      { timeout: 10_000 }
    );
  });

  describe("MP4", () => {
    it(
      "Video/Audio sync",
      async () => {
        const downloader = new Downloader(
          "33fPaNWvyzE" //https://www.youtube.com/watch?v=
        );
        await downloader.streamVideo();
      },
      { timeout: 50_000_000 }
    );
  });
});
