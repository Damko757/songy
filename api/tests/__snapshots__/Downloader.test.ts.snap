// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`FFMPEG MP3 Audio only 1`] = `
{
  "sigBytes": 32,
  "words": [
    675151534,
    812363979,
    874854964,
    -912879411,
    -1607286948,
    2067465579,
    -376983114,
    -723715869,
  ],
}
`;

exports[`FFMPEG MP3 Audio with Metadata 1`] = `
{
  "sigBytes": 32,
  "words": [
    -1590487131,
    80247706,
    376943452,
    1697193364,
    -1922663658,
    -1478818971,
    -105388837,
    1731139771,
  ],
}
`;
